docs/DesignDocs/OpteeIntegration.md

```diff
@@ -0,0 +1,321 @@
+by GlobalPlatform. A TA binary in the REE TA store thus looks like:
+```
+ Bin Hdr | ELF ( TA Hdr (.ta_info section) | .text | .bss | etc. )
+```
+As a result, one cannot load the resulting .ta file as a standard ELF binary
+unless one knows to skip past the TA binary header.
+For OP-TEE 3.6.0 (which the Open Enclave SDK currently uses), the TA binary
+header is in fact composed of several structures.
+
+First, there is the signed header (`shdr`): it contains a magic value [u32],
+the image type (i.e., Bootstrap TA or Encrypted TA) [u32], the image size
+(i.e., the size of the standard ELF binary [u32], the ID according to
+GlobalPlatform of the algorithm used to compute the TA's signature [u32],
+the length in bytes of the hash of some of the headers and the image [u16],
+the size of the signature of said hash [u16], the hash itself [u8[]], and
+the signature of the hash [u8[]].
+
+Second, there is the bootstrap TA header (`shdr_bootstrap_ta`): it contains
+the UUID of the TA [u8[]], and the TA version (used for rollback protection)
+[u32].
+
+Third, there is the encryption header (`shdr_encrypted_ta`): it contains the
+ID according to GlobalPlatform of the encryption algorithm [u32], the flags
+for said algorithm [u32], the size of its IV [u16], the size of its
+auth. tag [u16], the IV [u8[]], and the auth. tag [u8[]].
+
+```
+struct shdr {
+  uint32_t magic;
+  uint32_t img_type;
+  uint32_t img_size;
+  uint32_t algo;
+  uint16_t hash_size;
+  uint16_t sig_size;
+
+  /*
+   * Dynamically-sized arrays:
+   *
+   * uint8_t hash[hash_size];
+   * uint8_t sig[sig_size];
+   */
+};
+
+struct shdr_bootstrap_ta {
+  uint8_t uuid[sizeof(TEE_UUID)];
+  uint32_t ta_version;
+};
+
+struct shdr_encrypted_ta {
+  uint32_t enc_algo;
+  uint32_t flags;
+  uint16_t iv_size;
+  uint16_t tag_size;
+
+  /*
+   * Dynamically-sized arrays:
+   *
+   * uint8_t iv[iv_size];
+   * uint8_t tag[tag_size];
+   */
+};
+```
+
+Encrypting TAs is optional.
+For an encrypted TA, the hash is computed as follows:
+```
+H = SHA_256(shdr | shdr_bootstrap_ta | shdr_encrypted_ta | standard ELF)
+```
+
+For a cleartext TA, the hash is computed as follows:
+```
+H = SHA_256(shdr | shdr_bootstrap_ta | standard ELF)
+```
+
+The signature is computed as follows:
+```
+S = RSA_SSA_PKCS1_PSS_MGF1_SHA256(H)
+```
+
+The ciphertext is computed as follows:
+```
+C = AES_GCM(standard ELF)
+```
+
+For an encrypted TA, the final TA image is composed as follows:
+```
+TA = (shdr | H | S | shdr_bootstrap_ta | shdr_encrypted_ta | C)
+```
+
+For a cleartext TA, the final TA image is composed as follows:
+```
+TA = (shdr | H | S | shdr_bootstrap_ta | standard ELF)
+```
+
+For OP-TEE 3.6.0, `shdr_encrypted_ta` does not exist and there only exists
+the Bootstrap TA image type.
+
+When OP-TEE reads a TA binary
+from the REE TA store, the store takes care of performing signature
+verification and decryption.
+
+By the time a TA binary reaches the TA loader proper (i.e., the component
```

Comment:
Is the loader "proper" a specific terminology to OP-TEE?

Reply:
No, that's just my writing. I added "proper" to distinguish between the REE TA Store, which "loads" a binary from the non-secure filesystem, and the ELF loader that goes through the binary and readies it in memory for execution.

Reply:
added paragraph defining the term